File: src/net/liplum/animations/AniStateM.java
package net.liplum.animations;

import arc.util.Nullable;
import mindustry.gen.Building;
import mindustry.world.Block;

public class AniStateM<TBlock extends Block, TBuild extends Building> {
    private final TBlock block;
    private final TBuild build;
    private final AniConfig<TBlock, TBuild> config;
    private AniState<TBlock, TBuild> curState;
    @Nullable
    private ISwitchAniStateListener<TBlock, TBuild> switchAniStateListener;

    public AniStateM(AniConfig<TBlock, TBuild> config, TBlock block, TBuild build) {
        this.config = config;
        this.curState = config.getDefaultState();
        this.block = block;
        this.build = build;
    }

    public ISwitchAniStateListener<TBlock, TBuild> getSwitchAniStateListener() {
        return switchAniStateListener;
    }

    public void setSwitchAniStateListener(ISwitchAniStateListener<TBlock, TBuild> switchAniStateListener) {
        this.switchAniStateListener = switchAniStateListener;
    }

    public AniState<TBlock, TBuild> getCurState() {
        return curState;
    }

    public void setCurState(AniState<TBlock, TBuild> curState) {
        this.curState = curState;
    }

    public void drawBuilding() {
        curState.drawBuilding(this.block, this.build);
    }

    public void update() {
        for (AniState<TBlock, TBuild> to : config.getAllEntrances(curState)) {
            ITrigger<TBlock, TBuild> canEnter = config.getCanEnter(curState, to);
            if (canEnter != null && canEnter.canTrigger(this.block, this.build)) {
                if (switchAniStateListener != null) {
                    switchAniStateListener.OnSwitch(this.block, this.build, curState, to);
                }
                setCurState(to);
                return;
            }
        }
    }
}

File: mdt/src/net/liplum/mdt/animation/anis/AniStateM.kt
package net.liplum.mdt.animation.anis

import arc.graphics.g2d.Draw
import mindustry.gen.Building
import mindustry.world.Block
import net.liplum.mdt.render.RESET_CONTEXT

interface ISwitchAniStateListener<TBuild : Building> {
    fun onSwitch(build: TBuild, from: AniState<TBuild>, to: AniState<TBuild>)
}

open class AniStateM<TBuild : Building>(
    val config: AniConfig<TBuild>,
    val build: TBuild,
) {
    var transition: TransitionEffect = config.transition
    var transitionDuration: Float = config.transitionDuration
    var curState: AniState<TBuild> = config.defaultState!!
    var lastState: AniState<TBuild>? = null
    var switchAniStateListener: ISwitchAniStateListener<TBuild>? = null
    var onUpdate: (() -> Unit)? = null
        private set

    open fun onUpdate(onUpdate: () -> Unit): AniStateM<TBuild> {
        this.onUpdate = onUpdate
        return this
    }
    /**
     * For java
     */
    fun onUpdate(onUpdate: Runnable): AniStateM<TBuild> {
        onUpdate {
            onUpdate.run()
        }
        return this
    }

    var lastSwitchTime = 0f
    var curTime = 0f
    open fun spend(time: Float) {
        curTime += time
    }

    open fun drawBuilding() {
        val delta = curTime - lastSwitchTime
        val progress = (delta / transitionDuration).coerceIn(0f, 1f)
        transition(progress, {
            lastState?.drawBuilding(build)
        }, {
            curState.drawBuilding(build)
        })
        Draw.reset()
        RESET_CONTEXT()
    }

    open fun curOverwriteBlock(): Boolean = curState.isOverwriteBlock
    open fun update() {
        onUpdate?.invoke()
        for (to in config.getAllEntrances(curState)) {
            val canEnter = config.getCanEnter(curState, to)
            if (canEnter != null && canEnter(build)) {
                switchAniStateListener?.onSwitch(build, curState, to)
                lastState = curState
                curState = to
                lastSwitchTime = curTime
                return
            }
        }
    }
}